Feat/single tech/feature engineering on fabric br/setup env notebook

siliang-j-1225 opened the pull request:
Add optional setup environment notebook for simplified Azure resource setup.

# Type of PR
- Documentation changes
- Code changes

## Purpose
Add a deploy.ipynb notebook to assist setting up environment in feature engineering pipeline of Fabric.

## Does this introduce a breaking change? If yes, details on what can break
No.

## Author pre-publish checklist
<!-- Please check check before publishing PR using "x". Remove a column if it's not applicable. -->
- [ ] Added test to prove my fix is effective or new feature works
- [x] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps
Follow instructions in the notebook to create environment.

## Issues Closed or Referenced
No.

## Additional Information
Immigrated from this PR https://github.com/Azure-Samples/modern-data-warehouse-dataops/pull/653

promisinganuj commented:
@thurstonchen , @siliang-j-1225 , this is something new to me and I like it a lot. My concern is that it would add a bit of confusion to the person reading through the instructions. 

I would propose to just add a one-line footnote at the end of the "Required resources" section and put a link to the notebook there.

Few other comments:
- Move the "deploy.ipynb" to "src/infra/deploy-azure-resources.ipynb".
- Remove Step 2 and Step 3 completely from the notebook, these instructions are already present in the main README.md and we want people to use this notebook as a standalone artifact.
- For the following comments in the notebook:
```
run jupyter notebooks
authenticate with credential of the target tenant
```
We would either need to provide some more guidance about it or remove it altogether. For example, the reader would be wondering about which Azure service to use.
- In the notebook, please add all the details about what it does, how it does in the markdown so that it's self-contained. I would expect people to also use it as a reference point for such a unique deployment method.

siliang-j-1225 commented:
> @thurstonchen , @siliang-j-1225 , this is something new to me and I like it a lot. My concern is that it would add a bit of confusion to the person reading through the instructions.
> 
> I would propose to just add a one-line footnote at the end of the "Required resources" section and put a link to the notebook there. ***--> Removed the previous paragraph and added a link at the end of the "Required resources" section***
> 
> Few other comments:
> 
> * Move the "deploy.ipynb" to "src/infra/deploy-azure-resources.ipynb".  ***-->done***
> * Remove Step 2 and Step 3 completely from the notebook, these instructions are already present in the main README.md and we want people to use this notebook as a standalone artifact. ***-->done***
> * For the following comments in the notebook:
> 
> ```
> run jupyter notebooks
> authenticate with credential of the target tenant
> ```
> 
> We would either need to provide some more guidance about it or remove it altogether. ***-->decided to remove it altogether.*** For example, the reader would be wondering about which Azure service to use. ***-->added new contents in the beginning of the notebook shows which services will be created and what roles will be assigned***
> 
> * In the notebook, please add all the details about what it does, how it does in the markdown so that it's self-contained. I would expect people to also use it as a reference point for such a unique deployment method. ***-->I added some more details to markdown for each cell***
> * Also remove "Step 1. Create Azure Resources" header as it won't be required with the new change. ***-->done***

Hey @promisinganuj,  thanks for the informative comment! I have made some updates of the notebook and readme based on your comment. Please have a look later.

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/README.md:
```diff
@@ -102,6 +103,10 @@ As described above, the sample uses Microsoft Fabric as the data analytics platf
 
   For granting access to Microsoft Purview access, assign [Data curators](https://learn.microsoft.com/en-us/purview/how-to-create-and-manage-collections#roles) role in the Purview root collection to this SP. For details, refer to [Microsoft documentation](https://learn.microsoft.com/purview/how-to-create-and-manage-collections#add-role-assignments).
 
+
```
```suggestion
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/README.md:
```diff
@@ -102,6 +103,10 @@ As described above, the sample uses Microsoft Fabric as the data analytics platf
   For granting access to Microsoft Purview access, assign [Data curators](https://learn.microsoft.com/en-us/purview/how-to-create-and-manage-collections#roles) role in the Purview root collection to this SP. For details, refer to [Microsoft documentation](https://learn.microsoft.com/purview/how-to-create-and-manage-collections#add-role-assignments).
 
+
+#### [Optional] Notebook for Creating Above Resources
```
```suggestion
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/README.md:
```diff
@@ -102,6 +103,10 @@ As described above, the sample uses Microsoft Fabric as the data analytics platf
 
   For granting access to Microsoft Purview access, assign [Data curators](https://learn.microsoft.com/en-us/purview/how-to-create-and-manage-collections#roles) role in the Purview root collection to this SP. For details, refer to [Microsoft documentation](https://learn.microsoft.com/purview/how-to-create-and-manage-collections#add-role-assignments).
 
+
+#### [Optional] Notebook for Creating Above Resources
+To streamline your setup process, utilize the [deploy-azure-resources.ipynb](./src/infra/deploy-azure-resources.ipynb) notebook. This tool is designed to guide you through the creation of Azure resources mentioned above in an interactive and user-friendly way.
```
```suggestion
> Note: An optional (and novel) alternative for deploying the specified Azure resources involves utilizing the Fabric notebook. The [deploy-azure-resources.ipynb](./src/infra/deploy-azure-resources.ipynb) notebook can be used to interactively create the aforementioned resources. It utilizes Azure's Python SDK along with Azure CLI to facilitate the resource creation process.
```

promisinganuj replied:
```suggestion
> Note: *An optional (and novel) alternative for deploying the specified Azure resources involves utilizing the Fabric notebook. The [deploy-azure-resources.ipynb](./src/infra/deploy-azure-resources.ipynb) notebook can be used to interactively create the aforementioned resources. It utilizes Azure's Python SDK along with Azure CLI to facilitate the resource creation process.*
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+    "- Assign the Service Principal with following roles\n",
+    "  - `AzureML Data Scientist` role to the Azure ML Managed Feature Store\n",
+    "  - `Contributor` role to the Azure Purview Account\n",
+    "    - *Notice: this notebook can **not** help you to assign Service Principal with `Data curators` role of Azure Purview Account.* For manually assigning the `Data curators` role, please find more details in [README.md](../../README.md)"
```
```suggestion
    "    - Note: This notebook doesn't assign Purview's `Data Curator` role to the service principal. For manually assigning this role, refer to [Microsoft documentation](https://learn.microsoft.com/purview/how-to-create-and-manage-collections#add-role-assignments).
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+    "You will need to assign the Service Principal to the role `Data curators` in the Purview root collection manually. Check more details in [README.md](./README.md) and search for \"Data Lineage Setup\"."
+   ]
+  },
+  {
```
```suggestion
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
```suggestion
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+    "    --scope $purview_arm_id"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You will need to assign the Service Principal to the role `Data curators` in the Purview root collection manually. Check more details in [README.md](./README.md) and search for \"Data Lineage Setup\"."
+   ]
+  },
```
```suggestion
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+    "You will later need to setup the Fabric workspace Environment by uploading `./src/environment/sparkProperties.yaml` in one of the steps. The `sparkProperties.yaml` requires details from the created resources above.\n",
+    "\n",
+    "Run the following cell to get most of the values and replace the contents in the `sparkProperties.yaml`.  \n",
+    "*Notice: for `<fabric-tenant-name>` you still need to provide the information manually, which is able to obtain when later creating the Fabric workspace.*"
```
```suggestion
    "*Note: for `<fabric-tenant-name>` you still need to provide the information manually, which is able to obtain when later creating the Fabric workspace.*"
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/src/infra/deploy-azure-resources.ipynb:
```diff
@@ -0,0 +1,478 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You will later need to setup the Fabric workspace Environment by uploading `./src/environment/sparkProperties.yaml` in one of the steps. The `sparkProperties.yaml` requires details from the created resources above.\n",
+    "\n",
+    "Run the following cell to get most of the values and replace the contents in the `sparkProperties.yaml`.  \n",
```
```suggestion
    "You will later need to setup the Fabric workspace Environment by uploading [src/environment/sparkProperties.yml](./../environment/sparkProperties.yml) in one of the steps. The `sparkProperties.yml` requires the details of the resources created above. Run the following cell to get most of the values.\n",
```

promisinganuj commented on single_tech_samples/fabric/feature_engineering_on_fabric/README.md:
```diff
@@ -66,6 +66,7 @@ This subset contains data for the year 2022, and each month is available as a se
 - Access to an Azure subscription. You can sign up for a free trial by following [this link](https://azure.microsoft.com/en-us/free/).
 - Permissions to [create a service principal](https://learn.microsoft.com/entra/identity-platform/howto-create-service-principal-portal#register-an-application-with-microsoft-entra-id-and-create-a-service-principal) in Microsoft Entra ID.
 
+
```
```suggestion
```

promisinganuj commented:
Hi @siliang-j-1225 , thanks for the changes. I pushed few minor modifications and I am happy for this PR to be merged now. Well done!

promisinganuj reviewed (approved):
lgtm!
